worker/queue.go

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright 2021 Dgraph Labs, Inc. and Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package worker
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"math/rand"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/dgraph-io/dgraph/protos/pb"
+	"github.com/dgraph-io/dgraph/raftwal"
+	"github.com/dgraph-io/dgraph/x"
+	"github.com/dgraph-io/ristretto/z"
+	"github.com/golang/glog"
+)
+
+var (
+	Tasks tasks
+)
+
+func InitTasks() {
+	// #nosec G404: weak RNG
+	Tasks = tasks{
+		queue:  make(chan taskRequest, 16),
+		log:    z.NewTree(),
+		logMu:  new(sync.Mutex),
+		raftId: State.WALstore.Uint(raftwal.RaftId),
+		rng:    rand.New(rand.NewSource(time.Now().UnixNano())),
+	}
+	Tasks.deleteExpired()
+	Tasks.cancelQueued()
+	go Tasks.run()
+}
+
+type tasks struct {
+	// queue stores the full Protobuf request.
+	queue chan taskRequest
+	// log stores the timestamp, TaskKind, and TaskStatus.
+	log   *z.Tree
+	logMu *sync.Mutex
+
+	raftId uint64
+	rng    *rand.Rand
+}
+
+// Returns 0 if the task was not found.
+func (t tasks) Get(id uint64) TaskMeta {
+	if id == 0 || id == math.MaxUint64 {
+		return 0
+	}
+	t.logMu.Lock()
+	defer t.logMu.Unlock()
+	return TaskMeta(t.log.Get(id))
+}
+
+func (t tasks) set(id uint64, meta TaskMeta) {
+	t.logMu.Lock()
+	defer t.logMu.Unlock()
+	t.log.Set(id, uint64(meta))
+}
+
+// cancelQueuedTasks marks all queued tasks in the log as canceled.
+func (t tasks) cancelQueued() {
+	t.logMu.Lock()
+	defer t.logMu.Unlock()
+
+	t.log.IterateKV(func(id, val uint64) uint64 {
+		meta := TaskMeta(val)
+		if meta.Status() == TaskStatusQueued {
```

Comment:
What if the task was running? This should be TaskStatusQueued | TaskStatusRunning.